subsys/sensing/sensor_mgmt.c

```diff
@@ -0,0 +1,299 @@
+		sensor->conns[i].data = tmp_data;
+		tmp_data = (uint8_t *)tmp_data + get_sensor_sample_size_from_dt(dt, i);
+	}
+
+	__ASSERT(tmp_data == ((uint8_t *)sensor + total_size), "sensor memory assign error");
+
+	sensor->dt = dt;
+	sensor->dt->info->flags = sensor_ctx->register_info->flags;
+	sensor->dt->info->version = sensor_ctx->register_info->version;
+
+	LOG_INF("create sensor, sensor:%s, min_ri:%d(us)",
+		sensor->dt->dev->name, sensor->dt->info->minimal_interval);
+
+	sensor->interval = 0;
+	sensor->sensitivity_count = sensor_ctx->register_info->sensitivity_count;
+	__ASSERT(sensor->sensitivity_count <= CONFIG_SENSING_MAX_SENSITIVITY_COUNT,
+		 "sensitivity count:%d should not exceed MAX_SENSITIVITY_COUNT",
+		 sensor->sensitivity_count);
+	memset(sensor->sensitivity, 0x00, sizeof(sensor->sensitivity));
+
+	sensor->state = SENSING_SENSOR_STATE_OFFLINE;
+	sys_slist_init(&sensor->client_list);
+
+	sensor_ctx->priv_ptr = sensor;
+
+	return sensor;
+}
+
+static int sensing_init(void)
+{
+	struct sensing_context *ctx = &sensing_ctx;
+	struct sensing_sensor *sensor;
+	enum sensing_sensor_state state;
+	int ret = 0;
+	int i = 0;
+
+	LOG_INF("sensing init begin...");
+
+	if (ctx->sensing_initialized) {
+		LOG_INF("sensing is already initialized");
+		return 0;
+	}
+
+	if (ctx->sensor_num == 0) {
+		LOG_WRN("no sensor created by device tree yet");
+		return 0;
+	}
+
+	STRUCT_SECTION_FOREACH(sensing_dt_info, dt_info) {
+		sensor = create_sensor(dt_info);
+		if (!sensor) {
+			LOG_ERR("sensing init, create sensor error");
+			return -EINVAL;
+		}
+		ctx->sensors[i++] = sensor;
+	}
+
+	for_each_sensor(ctx, i, sensor) {
+		ret = init_sensor(sensor, sensor->dt->reporter_num);
+		if (ret) {
+			LOG_ERR("sensor:%s initial error", sensor->dt->dev->name);
+		}
+		state = (ret ? SENSING_SENSOR_STATE_OFFLINE : SENSING_SENSOR_STATE_READY);
+		ret = set_sensor_state(sensor, state);
+		if (ret) {
+			LOG_ERR("set sensor:%s state:%d error", sensor->dt->dev->name, state);
+		}
+		LOG_INF("sensing init, sensor:%s ret:%d", sensor->dt->dev->name, ret);
+	}
+
+	return ret;
+}
+
+int open_sensor(struct sensing_sensor *sensor, sensing_sensor_handle_t *handle)
+{
+	struct sensing_connection *conn;
+
+	*handle = NULL;
+	/* allocate struct sensing_connection *conn and conn data for application client */
+	conn = malloc(sizeof(*conn) + sensor->sample_size);
+	if (!conn) {
+		LOG_ERR("malloc memory for struct sensing_connection error");
+		return -ENOMEM;
+	}
+	conn->data = (uint8_t *)conn + sizeof(*conn);
+
+	/* create connection from sensor to application(client = NULL) */
+	init_connection(conn, sensor, NULL);
+
+	*handle = conn;
+
+	return 0;
+}
+
+int close_sensor(struct sensing_connection *conn)
+{
+	__ASSERT(conn && conn->source, "connection or reporter not be NULL");
+
+	__ASSERT(!conn->sink, "sensor derived from device tree cannot be closed");
+
+	sys_slist_find_and_remove(&conn->source->client_list, &conn->snode);
+
+	free(conn);
+
+	return 0;
+}
```

Comment:
Non-blocking : this pair of routines should validate that the application doesn't try to open a sensor twice, without closing.